balsam/platform/app_run/app_run.py

```diff
@@ -67,10 +68,23 @@ def get_num_ranks(self) -> int:
         return self._ranks_per_node * len(self._node_spec.node_ids)
 
     def get_cpus_per_rank(self) -> int:
-        cpu_per_rank = len(self._node_spec.cpu_ids[0]) // self._ranks_per_node
-        if not cpu_per_rank:
-            cpu_per_rank = max(1, int(self._threads_per_rank // self._threads_per_core))
-        return cpu_per_rank
+        # Get the list of cpus assigned to the job.  If it is a single node job, that is stored in
+        # the NodeSpec object.  If it is a multinode job, the cpu_ids assigned to NodeSpec is empty,
+        # so we will assume all cpus on a compute node are available to the job.  The list of cpus is
+        # just the list of cpus on the node in that case.
+        cpu_ids = self._node_spec.cpu_ids[0]
+        cpus_per_node = len(cpu_ids)
+        if not cpu_ids:
+            compute_node = ComputeNode(self._node_spec.node_ids[0], self._node_spec.hostnames[0])
```

Comment:
One thing to watch out for here is that `ComputeNode` will always have an empty list of `cpu_ids` since it's the generic base class.  Only at runtime do the launchers load the specific compute node class from the site configuration.  

I think it should be safe for the `NodeManager` to include the full set of CPU IDs in the `NodeSpec` of a multinode job.  This can be done by updating the method `NodeManager._assign_multi_node`.

That would keep the abstraction intact and avoid the need to pass an extra piece of information (which subclass of `ComputeNode`) from the launcher to the `AppRun`.  

https://github.com/argonne-lcf/balsam/blob/main/balsam/site/launcher/node_manager.py#L68

Reply:
I thought about doing that (modifying `_assign_multi_node`) but the way the NodeSpec object is structured, `cpu_ids` and `gpu_ids` will then be lists of a bunch of identical lists stored in memory.  That seemed a bit silly.  I did wonder why `cpu_ids` and `gpu_ids` have to be lists of lists?  They only time they contain non-empty lists in the single node case.